File: leetcode/LC/graph/1615_maximal_network_rank.py
"""
https://leetcode.com/problems/maximal-network-rank/discuss/1235181/C%2B%2B-O(N2)-and-O(N)-easy-to-understand
O(n^2) | O(n)    
"""

# take 1 1D vector to store road count, and one 2D vector to store connections

from typing import List

class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        maxrank = 0
        rank = [0 for i in range(n)]
        connections = set()
        
        for road in roads:
            connections.add((road[0], road[1]))
            rank[road[0]] += 1
            rank[road[1]] += 1
        
        for i in range(n):
            for j in range(i+1, n):
                currentrank = rank[i] + rank[j]
                if (i,j) in connections or (j, i) in connections:
                    currentrank -= 1
                maxrank = max(maxrank, currentrank)
        return maxrank
        

File: leetcode/LC1/graph/1615_maximal_network_rank.cpp
#include <climits>
#include <iostream>
using namespace std;
#include <vector>

// take 1 1D vector to store road count, and one 2D vector to store connections
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        int max_rank = 0;
        vector<int> v(n,0);
        vector<vector<int>> connect(n, vector<int>(n,0));
        
        for(int i = 0; i< roads.size(); i++){
        
            v[roads[i][0]]++;
            v[roads[i][1]]++;
            connect[roads[i][0]][roads[i][1]] = 1;
            connect[roads[i][1]][roads[i][0]] = 1;
        }
        
        for(int i = 0; i<n ;i++){
            for(int j = i+1; j < n; j++){
                max_rank = max(max_rank, v[i] + v[j] - connect[i][j]);
            }
        }
        return max_rank;
    }
};
